### DISCUSSION TITLE
Weird bug with charts

### DISCUSSION THREAD

**Discussion by THEkatinamicrowave** [CodenameCrew/CodenameEngine#299] [Q&A] [answered]
To 2SchoolHouse and maybe also Nex

I'm so confused
https://youtu.be/Vy8MejoWeaQ
if you find out why this happens, please tell me how to not make it happen

Sincerely,
-- hi

https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/cc7f2012-e4c3-4cdf-99cb-f1fddb92f14d

**Comment by r6915ee** [comment, marked as answer]
Generally, I'd say it would be good practice to instead have this set up as a setting within the mod itself to reduce the need of adding additional chart files.

Based on the way you have set up the mod (according to the video), it uses the assets folder, so by default you should have an `options.xml` file already located within the `data/config` folder. Much of it should include the basic types of settings, such as a float, though in your case something like this should add the family friendly option to the menu:

```xml
<!DOCTYPE codename-mod-options>
<!-- TEMPLATE FOR CUSTOM OPTIONS!!! -->
<menu name="Peter Griffin" desc="Modify mod options here.">
	<checkbox id="familyFriendly" name="Family friendly" />
</menu>
```

There are several types of options, though in your case a checkbox should work. It is a boolean option, allowing for either a true or false statement, just like a normal one.

Afterwards, if you want your option to be used on a script, you will need to surround the main code that should be affected by the option with an if statement. Something like this shall do if it doesn't remain when turned off:
```haxe
if (FlxG.save.data.familyFriendly) {
     // Enter your code here.
}
```

I've included a little section to explain why FlxG is necessary for the above.
<details><summary><h2>Why FlxG?</h2></summary>
<p>
FlxG is used commonly for save data. It is where save data variables are stored for use; in fact, one example of these kinds of variables is any custom option included in a Codename mod. Any custom option IS a variable as they are parsed and inserted to be put into use.

The ID that was included in the `options.xml` file is received to actually save it, hence why it's included in the condition in the first place. It is necessary to insert it in there, otherwise it likely will not work.
</p>
</details>

**Reply by THEkatinamicrowave** [reply]
This has been truly epic to discover. So now here's my main goal:

I have two vocal files
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/b5d7ba5e-e2a1-4038-9977-41b492bffe82)
what I want to happen is if Fambily Frienby is on, the chart vocals should have the vocal suffix of the Fambily Frienby one. I'll reply to this if I come back with nothing.

**Reply by THEkatinamicrowave** [reply]
I couldn't find out how to do it :(

**Reply by r6915ee** [reply]
> This has been truly epic to discover. So now here's my main goal:
> 
> I have two vocal files ![image](https://private-user-images.githubusercontent.com/89645987/331889901-b5d7ba5e-e2a1-4038-9977-41b492bffe82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYxNDQyMzUsIm5iZiI6MTcxNjE0MzkzNSwicGF0aCI6Ii84OTY0NTk4Ny8zMzE4ODk5MDEtYjVkN2JhNWUtZTJhMS00MDM4LTk5NzctNDFiNDkyYmZmZTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE5VDE4Mzg1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjZjJmMjU5ZDJhYTEyNjhkMDgzYTk4ZGU0YTY3MDM5NmRhNzRmYjc5OWI2NTcyY2FjOGY1ZTVhYmI1ZGY2NDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._VPy_kto8Tarses3LwKMhuT8NRRB8cF3lfNiIs1idbk) what I want to happen is if Fambily Frienby is on, the chart vocals should have the vocal suffix of the Fambily Frienby one. I'll reply to this if I come back with nothing.

I haven't tested this, but knowing how PlayState does something like this, you could perhaps use the line of code below in an if statement from the above. Just in case, make sure to import `funkin.game.PlayState` first and then add this to your if statement; I recommend adding it to one of the functions that is loaded on startup:

`vocals = FlxG.sound.load(Paths.voices(SONG.meta.name, "fambily-friemby"));`

`vocals` is an actual variable in PlayState, and is included within the Codename API documentation, used for the vocals. The function is included within PlayState itself, the only difference being that a difficulty is included.

**Reply by THEkatinamicrowave** [reply]
I script mostly with Roblox LUA so I hope this would work
```haxe
import FlxG.save.data.fambilyFrienby
import funkin.game.PlayState

if (FlxG.save.data.fambilyFrienby == true) {

    var vc = PlayState.strLine.vocals
    var sx = "fambily-frienby"

    vc = FlxG.sound.load(Paths.voices(SONG.meta.name, sx));
    
}
```
I'll update when I can

**Reply by r6915ee** [reply]
There are some things that should be fixed, so here is an updated script:

```haxe
import funkin.game.PlayState;

if (FlxG.save.data.fambilyFrienby == true) {

    var vc = PlayState.strLine.vocals;
    var sx = "fambily-frienby";

    vc = FlxG.sound.load(Paths.voices(SONG.meta.name, sx));
    
}
``` 

FlxG does not need to be imported, hence why it is excluded in the updated script. Any mentions of variables also require the semicolon, just like any normal function.

**Reply by THEkatinamicrowave** [reply]
epic to know (yeah that's not a thing in roblox lol)

**Reply by THEkatinamicrowave** [reply]
So I did a little bit of dicking around in the source and found Paths.hx. For some reason, the script wasn't working so I checked around to see what was going on and found this
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/e267da01-a1e5-4640-ad7e-75f7239d3f32)
Basically, there's 3 parameters to Paths.voices: song name, song difficulty, and prefix. The problem with this script we have is that we skipped over the song difficulty parameter. I'll try to fix this somehow (I'm too overconfident)

**Reply by THEkatinamicrowave** [reply]
oh yeah and here's the full code for the Paths.voices thing:

```haxe
inline static public function voices(song:String, difficulty:String = "normal", ?prefix:String = "")
	{
		var diff = getPath('songs/${song.toLowerCase()}/song/Voices$prefix-${difficulty.toLowerCase()}.$SOUND_EXT', null);
		return OpenFlAssets.exists(diff) ? diff : getPath('songs/${song.toLowerCase()}/song/Voices$prefix.$SOUND_EXT', null);
	}
```

**Reply by THEkatinamicrowave** [reply]
hoping this should work
```haxe
import funkin.game.PlayState

if (FlxG.save.data.fambilyFrienby == true) {

    var vc = PlayState.strLine.vocals;
    var px = "fambily-frienby";
    var blank = "";

    vc = FlxG.sound.load(Paths.voices(SONG.meta.name, px, blank));

}
```

**Reply by THEkatinamicrowave** [reply]
It didn't :(

**Reply by THEkatinamicrowave** [reply]
I found this in the console
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/3c076305-64d5-4fa3-bcd4-5aa5dbb18049)
oops

**Reply by THEkatinamicrowave** [reply]
I missed a semicolon again

**Reply by THEkatinamicrowave** [reply]
fixed semicolon
new error unlocked
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/db35debe-3fb8-4b01-a933-5ebe7ef198fd)

**Reply by THEkatinamicrowave** [reply]
I've officially made the decision to actually close this discussion and mark 2SchoolHouse's answer as the correct one and just assume I screwed up somewhere.

**Reply by r6915ee** [reply]
> So I did a little bit of dicking around in the source and found Paths.hx. For some reason, the script wasn't working so I checked around to see what was going on and found this ![image](https://private-user-images.githubusercontent.com/89645987/332081895-e267da01-a1e5-4640-ad7e-75f7239d3f32.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyNTgyODAsIm5iZiI6MTcxNjI1Nzk4MCwicGF0aCI6Ii84OTY0NTk4Ny8zMzIwODE4OTUtZTI2N2RhMDEtYTFlNS00NjQwLWFkN2UtNzVmNzIzOWQzZjMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDAyMTk0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1MzAxZGYyZWUyZGQ3N2I0ZTU1M2RiNGM4ZWIxYTFmYjNiMjE2MjRiYWQxOWJjYjFhZGViMzZkZmNlNDgyMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-sDypngfOlpOhCyZLE0qqp9xWcnFNW2Da5KMKdYCga8) Basically, there's 3 parameters to Paths.voices: song name, song difficulty, and prefix. The problem with this script we have is that we skipped over the song difficulty parameter. I'll try to fix this somehow (I'm too overconfident)

As far as I'm aware, the script should already have the difficulty parameter included, which in the latest revision that you included is `px`. The prefix one, to my knowledge, is an optional parameter, suggested by the question mark; it does not need to be set as it has a default option. Also, sorry for the late reply; I almost forgot about my Github notifications.
